astroid/bases.py

```diff
@@ -372,25 +382,46 @@ def infer_call_result(self, caller, context):
         on ``object.__new__`` will be of type ``object``,
         which is incorrect for the argument in general.
         If no context is given the ``object.__new__`` call argument will
-        correctly inferred except when inside a call that requires
+        be correctly inferred except when inside a call that requires
         the additional context (such as a classmethod) of the boundnode
         to determine which class the method was called from
         """
 
-        # If we're unbound method __new__ of builtin object, the result is an
+        # If we're unbound method __new__ of a builtin, the result is an
         # instance of the class given as first argument.
-        if (
-            self._proxied.name == "__new__"
-            and self._proxied.parent.frame(future=True).qname() == "builtins.object"
-        ):
-            if caller.args:
-                node_context = context.extra_context.get(caller.args[0])
-                infer = caller.args[0].infer(context=node_context)
-            else:
-                infer = []
-            return (Instance(x) if x is not Uninferable else x for x in infer)
+        if self._proxied.name == "__new__":
+            qname = self._proxied.parent.frame(future=True).qname()
+            # Avoid checking builtins.type: _infer_type_new_call() does more validation
+            if qname.startswith("builtins.") and qname != "builtins.type":
+                return self._infer_builtin_new(caller, context)
         return self._proxied.infer_call_result(caller, context)
 
+    def _infer_builtin_new(
+        self, caller: Call, context: InferenceContext
+    ) -> GenT[_BuiltinNewT, None, None]:
+        # pylint: disable-next=import-outside-toplevel; circular import
+        from astroid.nodes import Const, const_factory
+
+        if not caller.args:
+            yield from []
+            return
+        # Attempt to create a constant
+        if len(caller.args) > 1:
+            value = None
+            if isinstance(caller.args[1], Const):
+                value = caller.args[1].value
+            else:
+                inferred_arg = next(caller.args[1].infer(), None)
+                if isinstance(inferred_arg, Const):
+                    value = inferred_arg.value
+            if value is not None:
+                yield const_factory(value)
+
+        node_context = context.extra_context.get(caller.args[0])
+        infer = caller.args[0].infer(context=node_context)
+
+        yield from (Instance(x) if x is not Uninferable else x for x in infer)
```

Comment:
I didn't know if that was acceptable style to mix `yield` and `return ...` but happy to leave it as return 👍🏻

Reply:
As long as it works 🤷🏻‍♂️ Type checkers should complain otherwise, especially since you added type annotations.

Reply:
Interesting, I get a test failure with `return` that I haven't investigated.

Reply:
> Interesting, I get a test failure with `return` that I haven't investigated.

Might be my mistake. I overlooked that it's a generator method now. Thus `return` will raise a `StopIteration` with the generator object as argument, instead of returning the new generator. So `yield from` is probably correct.